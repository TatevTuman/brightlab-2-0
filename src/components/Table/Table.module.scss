.table {
  width: 100%;
  text-align: left;
  border-collapse: separate;
  margin-bottom: 32px;
  border-spacing: 0 4px;
}
