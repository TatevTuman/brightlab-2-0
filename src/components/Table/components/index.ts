export { default as TableBody } from './TableBody/TableBody'
export { default as TableHeader } from './TableHeader/TableHeader'
