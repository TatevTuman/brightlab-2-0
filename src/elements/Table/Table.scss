@import "static/assets/styles/mixins";

.grid-table {
  display: grid;
  border-top: 1px solid var(--purple);
  border-right: 1px solid var(--purple);

  &-column, &-cell {
    padding: toRem(8) toRem(4);
    border-bottom: 1px solid var(--purple);
    border-left: 1px solid var(--purple);
  }
}
