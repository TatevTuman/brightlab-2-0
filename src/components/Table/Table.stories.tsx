import React from 'react'
import Table from './Table'
import { Button } from '@elements'
import { TableColumnType } from '@types'

export default {
  title: 'Table',
  component: Table
}

export const Story = () => {
  const columns: TableColumnType[] = [
    {
      title: 'Имя',
      dataIndex: 'name',
      key: 'name'
    },
    {
      title: 'Фамилия',
      dataIndex: 'surname',
      key: 'surname'
    },
    {
      title: 'Возраст',
      dataIndex: 'age',
      key: 'age'
    },
    {
      title: '',
      dataIndex: 'info',
      key: 'info',
      render(cell, row) {
        const info = row.name + ' ' + row.surname + ' ' + row.age
        return (
          <Button type={'secondary'} size={'md'} onClick={() => alert(info)}>
            Get Info
          </Button>
        )
      }
    }
  ]

  const rows = new Array(10).fill(0).map(() => ({
    name: 'John',
    surname: 'Doe',
    age: '34'
  }))

  return (
    <div className="story">
      <Table columns={columns} rows={rows} handleRowClick={() => alert('Row clicked!')} isRowIndex />
    </div>
  )
}
