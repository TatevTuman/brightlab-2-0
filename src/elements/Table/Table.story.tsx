import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import Table, { TableProps } from './Table'
import { storybook } from '@utils'
import { AnyObject } from '@types'

interface TableStoryTemplateProps extends TableProps<AnyObject> {}

export default {
  title: 'Elements/Table',
  argTypes: {
    columns: { control: 'object', name: 'Columns' },
    data: { control: 'object', name: 'Data' }
  }
} as Meta

const TableStoryTemplate: Story<TableStoryTemplateProps> = args => {
  return <Table {...args} />
}

const columns = [
  {
    key: 'column1',
    title: 'Column1'
  },
  {
    key: 'column2',
    title: 'Column2'
  },
  {
    key: 'column3',
    title: 'Column3'
  },
  {
    key: 'column4',
    title: 'Column4'
  },
  {
    key: 'column5',
    title: 'Column5'
  },
  {
    key: 'column6',
    title: 'Column6'
  }
]

const data = new Array(10).fill(0).map((_, index) => {
  return columns.reduce<{ [x: string]: string }>((acc, column) => {
    return {
      ...acc,
      [column.key]: 'Lorem ipsum dolores sit amet'
    }
  }, {})
})

TableStoryTemplate.args = {
  columns,
  data
}

export const TableStory = TableStoryTemplate.bind({})
TableStory.args = {
  ...TableStoryTemplate.args
}
