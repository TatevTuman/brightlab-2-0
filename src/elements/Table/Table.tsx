import React, { CSSProperties, memo } from 'react'
import { get } from 'lodash'
import './Table.scss'

export type TableColumnType<T> = {
  key: string
  title: string
  style?: CSSProperties
  render?(field: any, item: T): JSX.Element
}

export type TableCellType = {
  disabled?: boolean
}

export interface TableProps<T> {
  columns: TableColumnType<T>[]
  data: T[]
}

const TableColumns = <T extends TableCellType>(props: { columns: TableColumnType<T>[] }) => {
  const { columns } = props

  return (
    <>
      {columns.map(columns => {
        const { key, title, style } = columns

        return (
          <div key={`column-${key}`} className={'grid-table-column'} style={style}>
            {title}
          </div>
        )
      })}
    </>
  )
}

const TableRows = <T extends TableCellType>(props: TableProps<T>) => {
  const { columns, data } = props

  return (
    <>
      {data.map((item: T, index: number) => {
        return columns.map(column => {
          const { key } = column

          const itemData = get(item, key, null)
          const render = column.render ? column.render(itemData, item) : itemData

          return (
            <div key={`cell-${key}-${index}`} className={'grid-table-cell'} data-disabled={item.disabled}>
              {render}
            </div>
          )
        })
      })}
    </>
  )
}

const Table = <T extends TableCellType>(props: TableProps<T>) => {
  const { columns, data } = props

  return (
    <div className={'grid-table'} style={{ gridTemplateColumns: `repeat(${columns.length}, 1fr)` }}>
      <TableColumns<T> columns={columns} />
      <TableRows<T> columns={columns} data={data} />
    </div>
  )
}

export default memo(Table)
