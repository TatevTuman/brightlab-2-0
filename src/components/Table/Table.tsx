import React, { memo } from 'react'
import { TableColumnType, TableRowType } from '@types'
import { TableHead, TableBody } from './components'
import './Table.scss'

interface TableProps {
  columns: TableColumnType[]
  rows: TableRowType[]
  filters?: { field: string; handleFilter: (...args: any) => boolean }[]
  handleRowClick?: (row: TableRowType) => void
  isRowIndex?: boolean
  minWidth?: string
}

const Table: React.FC<TableProps> = props => {
  const { columns, rows, filters, handleRowClick, isRowIndex, minWidth } = props

  return (
    <table className={'table'} style={{ minWidth }}>
      <TableHead columns={columns} isRowIndex={isRowIndex} />
      <TableBody
        rows={rows}
        columns={columns}
        filters={filters}
        handleRowClick={handleRowClick}
        isRowIndex={isRowIndex}
      />
    </table>
  )
}

export default memo(Table)
