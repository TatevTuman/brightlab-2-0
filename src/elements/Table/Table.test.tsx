import React from 'react'
import { render, cleanup, waitFor } from '@testing-library/react'
import Table, { TableProps } from './Table'
import userEvent from '@testing-library/user-event'

const columns = [
  {
    key: 'column1',
    title: 'Column1'
  },
  {
    key: 'column2',
    title: 'Column2'
  },
  {
    key: 'column3',
    title: 'Column3'
  },
  {
    key: 'column4',
    title: 'Column4'
  },
  {
    key: 'column5',
    title: 'Column5'
  },
  {
    key: 'column6',
    title: 'Column6'
  }
]

const data = new Array(10).fill(0).map((_, index) => {
  return columns.reduce<{ [x: string]: string }>((acc, column) => {
    return {
      ...acc,
      [column.key]: column.title + index
    }
  }, {})
})

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Table', () => {
  const props = { columns, data }

  it('renders correctly', async () => {
    const { container } = render(<Table {...props} />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
  })

  it('renders columns correctly', () => {
    const { getByText } = render(<Table {...props} />)

    columns.forEach((column, index) => {
      const { title } = column

      if (title) {
        expect(getByText(title + index)).toBeInTheDocument()
      }
    })
  })

  it('renders cells correctly', () => {
    const { getByText } = render(<Table {...props} />)

    data.forEach(item => {
      columns.forEach(column => {
        expect(getByText(item[column.key])).toBeInTheDocument()
      })
    })
  })

  // it('handles onRowClick', () => {
  //   const handleRowClick = jest.fn()
  //
  //   const { container } = render(<Table {...props} handleRowClick={handleRowClick} />)
  //   const disabledRowElement = container.querySelector('.table-body-tr[data-disabled=true]')
  //   const enabledRowElement = container.querySelector('.table-body-tr[data-disabled=false]')
  //
  //   userEvent.click(disabledRowElement!)
  //   expect(handleRowClick).toHaveBeenCalledTimes(0)
  //
  //   userEvent.click(enabledRowElement!)
  //   expect(handleRowClick).toHaveBeenCalledTimes(1)
  // })
  //
  // it('renders filters correctly', () => {
  //   const filters = [
  //     {
  //       field: 'age',
  //       handleFilter: jest.fn((age: number) => {
  //         return +age >= 5
  //       })
  //     }
  //   ]
  //
  //   const { container, rerender } = render(<Table {...props} />)
  //
  //   let rowElements = container.querySelectorAll('.table-body-tr')
  //   expect(rowElements).toHaveLength(10)
  //
  //   rerender(<Table {...props} filters={filters} />)
  //
  //   rowElements = container.querySelectorAll('.table-body-tr')
  //   expect(rowElements).toHaveLength(5)
  //   expect(filters[0].handleFilter).toHaveBeenCalledTimes(rows.length)
  // })
})
