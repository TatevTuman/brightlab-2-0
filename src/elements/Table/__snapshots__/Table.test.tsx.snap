// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
<div>
  <div
    class="grid-table"
    style="grid-template-columns: repeat(6, 1fr);"
  >
    <div
      class="grid-table-column"
    >
      Column1
    </div>
    <div
      class="grid-table-column"
    >
      Column2
    </div>
    <div
      class="grid-table-column"
    >
      Column3
    </div>
    <div
      class="grid-table-column"
    >
      Column4
    </div>
    <div
      class="grid-table-column"
    >
      Column5
    </div>
    <div
      class="grid-table-column"
    >
      Column6
    </div>
    <div
      class="grid-table-cell"
    >
      Column10
    </div>
    <div
      class="grid-table-cell"
    >
      Column20
    </div>
    <div
      class="grid-table-cell"
    >
      Column30
    </div>
    <div
      class="grid-table-cell"
    >
      Column40
    </div>
    <div
      class="grid-table-cell"
    >
      Column50
    </div>
    <div
      class="grid-table-cell"
    >
      Column60
    </div>
    <div
      class="grid-table-cell"
    >
      Column11
    </div>
    <div
      class="grid-table-cell"
    >
      Column21
    </div>
    <div
      class="grid-table-cell"
    >
      Column31
    </div>
    <div
      class="grid-table-cell"
    >
      Column41
    </div>
    <div
      class="grid-table-cell"
    >
      Column51
    </div>
    <div
      class="grid-table-cell"
    >
      Column61
    </div>
    <div
      class="grid-table-cell"
    >
      Column12
    </div>
    <div
      class="grid-table-cell"
    >
      Column22
    </div>
    <div
      class="grid-table-cell"
    >
      Column32
    </div>
    <div
      class="grid-table-cell"
    >
      Column42
    </div>
    <div
      class="grid-table-cell"
    >
      Column52
    </div>
    <div
      class="grid-table-cell"
    >
      Column62
    </div>
    <div
      class="grid-table-cell"
    >
      Column13
    </div>
    <div
      class="grid-table-cell"
    >
      Column23
    </div>
    <div
      class="grid-table-cell"
    >
      Column33
    </div>
    <div
      class="grid-table-cell"
    >
      Column43
    </div>
    <div
      class="grid-table-cell"
    >
      Column53
    </div>
    <div
      class="grid-table-cell"
    >
      Column63
    </div>
    <div
      class="grid-table-cell"
    >
      Column14
    </div>
    <div
      class="grid-table-cell"
    >
      Column24
    </div>
    <div
      class="grid-table-cell"
    >
      Column34
    </div>
    <div
      class="grid-table-cell"
    >
      Column44
    </div>
    <div
      class="grid-table-cell"
    >
      Column54
    </div>
    <div
      class="grid-table-cell"
    >
      Column64
    </div>
    <div
      class="grid-table-cell"
    >
      Column15
    </div>
    <div
      class="grid-table-cell"
    >
      Column25
    </div>
    <div
      class="grid-table-cell"
    >
      Column35
    </div>
    <div
      class="grid-table-cell"
    >
      Column45
    </div>
    <div
      class="grid-table-cell"
    >
      Column55
    </div>
    <div
      class="grid-table-cell"
    >
      Column65
    </div>
    <div
      class="grid-table-cell"
    >
      Column16
    </div>
    <div
      class="grid-table-cell"
    >
      Column26
    </div>
    <div
      class="grid-table-cell"
    >
      Column36
    </div>
    <div
      class="grid-table-cell"
    >
      Column46
    </div>
    <div
      class="grid-table-cell"
    >
      Column56
    </div>
    <div
      class="grid-table-cell"
    >
      Column66
    </div>
    <div
      class="grid-table-cell"
    >
      Column17
    </div>
    <div
      class="grid-table-cell"
    >
      Column27
    </div>
    <div
      class="grid-table-cell"
    >
      Column37
    </div>
    <div
      class="grid-table-cell"
    >
      Column47
    </div>
    <div
      class="grid-table-cell"
    >
      Column57
    </div>
    <div
      class="grid-table-cell"
    >
      Column67
    </div>
    <div
      class="grid-table-cell"
    >
      Column18
    </div>
    <div
      class="grid-table-cell"
    >
      Column28
    </div>
    <div
      class="grid-table-cell"
    >
      Column38
    </div>
    <div
      class="grid-table-cell"
    >
      Column48
    </div>
    <div
      class="grid-table-cell"
    >
      Column58
    </div>
    <div
      class="grid-table-cell"
    >
      Column68
    </div>
    <div
      class="grid-table-cell"
    >
      Column19
    </div>
    <div
      class="grid-table-cell"
    >
      Column29
    </div>
    <div
      class="grid-table-cell"
    >
      Column39
    </div>
    <div
      class="grid-table-cell"
    >
      Column49
    </div>
    <div
      class="grid-table-cell"
    >
      Column59
    </div>
    <div
      class="grid-table-cell"
    >
      Column69
    </div>
  </div>
</div>
`;
